[Alek Kwek] iP

AlekKwek18 opened the pull request:
# SilverWolf Bot

>  “*sigh* I wish this universe had a few more things to do. Don't you get bored running across the map every day?” – Silver Wolf ([source](https://honkai-star-rail.fandom.com/wiki/Silver_Wolf/Voice-Overs))

SilverWolf bot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/AlekKwek18/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

 - [x] Managing tasks
 - [x] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

AlekKwek18 committed:
Rename the file, added greet and exit

AlekKwek18 committed:
added echo commands and will end the program if user types the command bye

AlekKwek18 committed:
added the ability to read and store user input and print them as list. in addition, logo is added

AlekKwek18 committed:
created the Task class and refactored the code to use Task instead of String

AlekKwek18 committed:
added the ability to mark tasks as done and the ability to change the status back to not done.

AlekKwek18 committed:
Added ToDo as a subclass to Task and refactored the code base to use ToDo class over Task class

AlekKwek18 committed:
added Deadline class which is a subclass for Todo class. Added type behaviour to every class. Added todo as a command

AlekKwek18 committed:
rebranded with new icon and new end text

AlekKwek18 committed:
changed the output for mark, unmarked and list so that it shows the type of task

AlekKwek18 committed:
added toString to Task and its subclass to enable polymorphism. refactored the main code based to take advantage of polymorphism. Fixed minor formatting issues. Lastly, added the deadline functionality

AlekKwek18 committed:
added the functionality where users can add events now

AlekKwek18 committed:
refactored text-ui-test

AlekKwek18 committed:
added a few more testcases

AlekKwek18 committed:
refactored the inputs to have its own methods

AlekKwek18 committed:
added exceptions to handle errors

AlekKwek18 committed:
added delete feature and updated the testcase

AlekKwek18 committed:
changed task to abstract class and added enums

AlekKwek18 committed:
refactored the code such that event, deadline, and todo is a subclass of task which will conform to LSP.

AlekKwek18 committed:
added descriptive header for classes and methods

AlekKwek18 committed:
Add new toFileString method for file format

The code lacks a method for converting task data into a format suitable
for file storage. This absence hinders future development for persisting
tasks to a file.

To support file read and write operations, a method is required to
consistently represent tasks as strings that can be easily saved
and retrieved.

Introduce a new `toFileString` method that formats task data as
"TaskType | StatusIcon | Description | From-To" depending on the
Task type.
This format will be used for file storage.

This structure ensures that all relevant task details are included
in a consistent format, paving the way for reliable file I/O
operations in future development.

This is an initial step toward implementing task persistence,
and further changes will involve handling file read/write operations.

AlekKwek18 committed:
Add file read and write functionality

Currently, tasks are not saved or loaded from a file.

Implement file reading and writing to persist task data.

This ensures tasks are saved and restored between sessions.

AlekKwek18 committed:
Add ./data/silverwolf.txt to .gitignore

Prevent tracking of the data file used for storing runtime information.

This change avoids including dynamically generated
or user-specific data in version control.

AlekKwek18 committed:
Add empty file content check in parseTask method

Currently, the chatbot will throw an exception if the file is
empty.

Updated `parseTask` method to handle empty file cases gracefully.

Improved robustness by ensuring tasks are
only parsed if file content is present.

AlekKwek18 committed:
Add parseDateTime for date format validation

The chatbot currently cannot correctly interpret date formats
and simply accepts any input without validation.

This behavior can lead to errors and confusion
when users input dates in various formats,
expecting the chatbot to understand them.

Implement the parseDateTime method, which attempts to parse the input
against a list of predefined date and time formats.
If none of the formats match, it throws a SilverWolfException,
prompting the user to check their input.

By validating and parsing date formats, the chatbot becomes more robust
and user-friendly, preventing incorrect inputs and
providing clear feedback when issues arise.

The change improves overall functionality and ensures that
date inputs are handled consistently across the application. In
addition, test cases has been updated to take into account of the
new date format.

AlekKwek18 committed:
Remove unused LocalTime import

Currently, the java.time.LocalTime import is included
in the codebase, but it is not being used anywhere.

This import needs to be removed to clean up the code
and avoid confusion about unnecessary dependencies.

Remove the LocalTime import to maintain a clean and efficient codebase.

This is done to follow best practices of only including relevant
imports, improving code readability and maintainability.

AlekKwek18 committed:
Refactor chatbot for improved OOP design

The chatbot does not fully adhere to Object-Oriented
Programming principles.

The current code structure may be difficult to maintain,
extend, and understand.

Refactor the code to better utilize OOP concepts
such as encapsulation, inheritance, and polymorphism.

Adhering to OOP principles makes the codebase more modular,
scalable, and easier to manage.

Improved code structure facilitates future enhancements and debugging.

AlekKwek18 committed:
Refactor codebase into packages for improved organization

The codebase has all classes in a single package,
making it difficult to navigate and maintain.

Organizing classes into packages improves code readability and
maintainability. It also simplifies the management of
related functionalities and dependencies.

Action taken:
- Create `silverwolf.command` package and move command-related classes
(AddCommand, DeleteCommand, UnmarkCommand) into it.
- Create `silverwolf.parser` package and move parsing classes (
Parser, DateTimeParser) into it.
- Create `silverwolf.ui` package for UI-related classes
and `silverwolf.storage`
 package for storage-related classes.
- Introduce `silverwolf.task` package for task management classes
 (Task, TaskList).
- Create `silverwolf.exception` package and move any custom
exceptions into the package

This organization clarifies the project structure and helps in scaling
the project as more functionalities are added.
Each package now contains related classes,
enhancing modularity and reducing potential for code conflicts.

No changes to class functionality,
only restructuring for better code management.

AlekKwek18 committed:
Merge branch 'add-gradle-support' of https://github.com/AlekKwek18/ip

AlekKwek18 committed:
Set up Gradle for project build automation

Currently, the SilverWolf project is built and run
manually without an automated build system.

Integrate Gradle into the project by merging the `add-gradle-support`
branch into `master`. Configure the `build.gradle` file to support
building and running the SilverWolf project using Gradle.

Gradle is a powerful and flexible build automation tool that
simplifies dependency management, compilation, and execution processes.
Setting it up early in the project helps streamline development.

This setup enables automated builds and prepares the project for
future enhancements, such as automated testing and
deployment processes using Gradle.

AlekKwek18 committed:
Add JUnit tests for TaskList and Storage classes

The `Todo`,`TaskList` and `Storage` classes lack
comprehensive unit tests.

Ensuring the functionality of these classes through tests
is crucial for maintaining reliability and catching bugs early.

Implement unit tests for `TaskList` to verify task management operations
such as addition, retrieval, and size management.
Add tests for `Storage` to check saving and loading functionality,
handle empty files, and manage invalid paths.

Testing edge cases and typical use cases ensures that the classes
perform as expected under various conditions.
Improving comments provides clarity and facilitates easier maintenance.

Tests are designed to cover a broad range of scenarios,
including error handling and boundary conditions.

AlekKwek18 committed:
Package the application as an executable JAR file

The application is not yet distributed as an executable JAR file,
making it less convenient for users to run the application.

Packaging the application as a JAR file simplifies
 distribution and execution for end-users, providing a single,
easily runnable file.

Package the app into an executable JAR file following the provided
guidelines. Ensure that the JAR file can be run from a command line
with `java -jar "{filename}.jar"`. Make necessary adjustments to
the project setup to support JAR packaging.

Creating an executable JAR file streamlines deployment and
execution processes, making it easier for users to run the
application without additional setup.

The JAR file will not be committed to the Git repository.

AlekKwek18 committed:
Add find command to search tasks by keyword

Currently, the task management system lacks a feature for
searching tasks based on keywords.
This limits the ability to quickly locate specific
 tasks within a large list.

Implement the `execute` method in the `FindCommand` class to
enable users to search for tasks by a specified keyword.
The method retrieves tasks matching the keyword from the
`TaskList` and displays them using the `Ui`. The `Storage` parameter
 is included for method signature consistency
but is not used in this implementation.

This change provides a way for users to efficiently
find tasks by their descriptions, improving
overall task management and usability.

No changes to the `Storage` class were necessary for this functionality.

AlekKwek18 committed:
Add JavaDoc comments for code documentation

Enhance the code documentation by updating JavaDoc comments.

Provide comprehensive comments for most non-private classes and
methods, and for
 non-trivial private methods. This update improves code clarity
and maintainability by ensuring that the purpose and functionality
 of various components are clearly documented.

No changes to code functionality were made; only updates to
documentation were performed to facilitate better understanding
 and future maintenance.

AlekKwek18 committed:
Fix codebase to comply with Checkstyle rules

The codebase contains multiple Checkstyle violations,
including issues with import order, missing newlines,
inconsistent formatting, and other style-related problems.

These violations cause build failures and reduce code readability,
making it harder to maintain and collaborate on the project.

Address and fix all Checkstyle violations,
including reordering imports,
adding missing newlines, correcting formatting issues,
and ensuring code adheres to the project's style guidelines.

Ensures compliance with Checkstyle rules,
leading to successful builds, improved code consistency,
and easier maintenance.

AlekKwek18 committed:
Merge branch 'branch-A-CodingStandard'

AlekKwek18 committed:
Add tests for `findTasks` method in `TaskList`

The `findTasks` method in the `TaskList` class
lacks adequate testing for finding list.

To ensure the method correctly searches for tasks
based on keywords and handles edge cases properly.

- Verify search results when tasks match the keyword.
- Confirm behavior when no tasks match the keyword.
- Test partial keyword matches.
- Check functionality with an empty keyword.

Adding comprehensive tests will validate the correctness
of the search functionality and improve reliability.

The tests will help in identifying issues early and maintaining
the robustness of the search feature as the codebase evolves.

AlekKwek18 committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

# Conflicts:
#	build.gradle
#	src/main/java/silverwolf/task/TaskList.java
#	src/main/java/silverwolf/ui/Ui.java

AlekKwek18 committed:
Merge branch 'branch-Level-9'

AlekKwek18 committed:
Minor bug fixes

Currently, the Ui class is missing the List util which is
needed to search list.

Add import java.util.List

AlekKwek18 committed:
Update chat bot guide

Currently, the guide is not updated to the latest version. This may
cause confusion of users using the chat bot.

Update the guide with the full list of commands.

In this way, users will understand how to use the chat bot.

Fix minor bugs

AlekKwek18 committed:
Resolve Checkstyle violations

Checkstyle errors are present due to improper formatting,
variable naming inconsistencies, and other style violations.

These violations prevent successful builds and reduce code readability,
making it harder to maintain and collaborate.

Correct formatting, apply consistent naming conventions,
and adhere to Checkstyle rules across the codebase.

Following Checkstyle ensures consistency in code quality,
improves maintainability, and aligns with project standards.

No functionality changes are introduced; this commit solely
addresses code style compliance.

PresleyChew commented on src/main/java/silverwolf/SilverWolf.java:
```diff
@@ -0,0 +1,73 @@
+ * The main class for the SilverWolf chatbot application.
+ * It initializes the necessary components and runs the chatbot loop.
+ */
+public class SilverWolf {
```
Good naming convention, followed PascalCase 👍🏻

PresleyChew commented on src/main/java/silverwolf/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @return false by default, indicating that the application should not exit after this command.
+     */
+    public boolean isExit() {
```
Boolean method sound like booleans, good job! 💯

PresleyChew commented on src/main/java/silverwolf/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,68 @@
+package silverwolf.parser;
+
+import java.time.LocalDateTime;
```
Did not use generics while importing, nice !

PresleyChew reviewed (commented):
LGTM! Nothing to nit aboout! 😄

FYL2003 commented on src/main/java/silverwolf/SilverWolf.java:
```diff
@@ -0,0 +1,73 @@
+
+    // constants for directory and file paths
+    public static final String DIRECTORY_PATH = "data/";
+    public static final String FILE_PATH = DIRECTORY_PATH + "silverWolf.txt";
```
I like how you use SCREAMING_SNAKE_CASE for final variables

FYL2003 commented on src/main/java/silverwolf/storage/Storage.java:
```diff
@@ -0,0 +1,116 @@
+public class Storage {
+    private final String directoryPath;
+    private final String filePath;
+    private final String fullPath;
```
I like how you differentiate those variables, sounds clear to me

FYL2003 commented on src/main/java/silverwolf/task/Task.java:
```diff
@@ -0,0 +1,131 @@
+public abstract class Task {
+    protected TaskType type;
+    private String description;
+    private boolean isDone;
```
This sounds very boolean, and it looks good to me

FYL2003 commented on src/main/java/silverwolf/ui/Ui.java:
```diff
@@ -0,0 +1,161 @@
+    /**
+     * Displays a goodbye message to the user when the application exits.
+     */
+    public void showGoodbye() {
```
Maybe it would be better to change it to sayGoodbye?

FYL2003 commented on src/main/java/silverwolf/storage/Storage.java:
```diff
@@ -0,0 +1,116 @@
+                writer.newLine();
+            }
+            File toBeDeleted = new File(filePath);
+            toBeDeleted.delete();
```
The naming of the file clearly shows the purpose, nice!

FYL2003 reviewed (commented):
I think there is no problem, looks good to me

AlekKwek18 committed:
Migrate from CLI to JavaFX GUI

Currently, the application relies on a command-line interface,
limiting user interaction and usability.

This needs to change to improve the user experience and provide a
more intuitive, graphical way to interact with the app.

Refactor the application by replacing the command-line interface with a
JavaFX-based GUI. Implement FXML to manage layout and
event handling for user input.

JavaFX is chosen for its flexibility and ease of integration with
the existing codebase. FXML allows for better
separation of concerns between UI and logic.

Also ensure resource paths for FXML are correctly resolved,
and adjust controller methods to handle GUI events.

AlekKwek18 committed:
Merge branch 'add-gui'

AlekKwek18 commented:
# SilverWolf Bot

>  “*sigh* I wish this universe had a few more things to do. Don't you get bored running across the map every day?” – Silver Wolf ([source](https://honkai-star-rail.fandom.com/wiki/Silver_Wolf/Voice-Overs))

SilverWolf bot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/AlekKwek18/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

 - [X] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

AlekKwek18 committed:
Add Assertions To Validate Inputs

The command parsing currently does not enforce input validation
through assertions, which could lead to unexpected behaviour when
handling invalid or missing arguments.

Assertions have been added to ensure that key values like the command
word, task descriptions, and date/time information are not null or
empty.

This change improves error detection early in the process and helps
to prevent the program from executing invalid commands.

AlekKwek18 committed:
Merge pull request #1 from AlekKwek18/branch-A-Assertions

Add Assertions To Validate Inputs

AlekKwek18 committed:
Improve Code Quality

Enhance the code base by adding input validation in the parser
class. Fixed a type in the variable name.

This ensure the best coding practises are met.

AlekKwek18 committed:
Merge branch 'master' into branch-A-CodeQuality

AlekKwek18 committed:
Merge pull request #2 from AlekKwek18/branch-A-CodeQuality

Branch a code quality

AlekKwek18 committed:
Implement flexible command syntax and aliases

Enhance the command parsing functionality to
support friendlier and more flexible syntax.

Key changes include:
- Added support for shorter aliases for commands (e.g., "t" for "todo").
- Introduced user-defined aliases for greater customization.
- Allowed commands to be less rigid in their syntax and order.

These improvements aim to make the chatbot interface more intuitive
and user-friendly, allowing users to interact with the
system in a more natural and flexible manner.

AlekKwek18 committed:
Merge branch 'add-friendlier-syntax'

AlekKwek18 committed:
Enhance User Guide With Image And Task Features

The current user guide lacks a visual component and detailed
descriptions of key features, which can make it harder
for users to follow.

Add a picture and enhance the guide with descriptions,
including an introduction to the Silver Wolf bot for task tracking.
Expand on task types, main functions, and friendly
syntax for commands like deadline (d), event (e), and todo (t).

This improves user engagement and makes the guide more accessible and
easier to understand for users with short-term memory challenges.

AlekKwek18 committed:
Merge branch 'improve-user-guide'

AlekKwek18 committed:
Resolve minor issues in the user guide

The user guide currently contains minor typos, unclear instructions,
and outdated information, which could confuse users.

These issues need to be addressed to ensure users can easily
understand the guide and follow instructions without confusion.

Correct typos, rephrase unclear sections, and update outdated
information to reflect the current state of the product.

This approach improves readability and
maintains up-to-date content for users.

No major changes to the structure were needed,
so the guide retains its original format
while improving the overall quality.

AlekKwek18 committed:
Merge branch 'fix-user-guide'

AlekKwek18 committed:
Fix exception message and checkstyle

AlekKwek18 committed:
Merge branch 'minor-bug-fix'

AlekKwek18 committed:
Merge branch 'master' into minor-bug-fix

AlekKwek18 committed:
Fix mistake in exception message

AlekKwek18 committed:
Merge branch 'minor-bug-fix'

AlekKwek18 committed:
Fix friendlier syntax

AlekKwek18 committed:
Merge branch 'minor-bug-fix'

AlekKwek18 committed:
Add guide to run the application

AlekKwek18 committed:
Fix Storage Write and Delete Command

Currently, the delete command will throw an uncaught exception if the
user enters a index that is not in the task array. In addition,
Storage class will create two text file instead of one text file.

Fix the bug on the Storage class which will now create just one text
file. The delete command only has a index checker to catch any
invalid index.

This ensures that the application is bug free for the user.

AlekKwek18 committed:
Merge branch 'fix-write-issue'

AlekKwek18 committed:
Update Ui.png

AlekKwek18 committed:
Merge branch 'update-ui-png'
